File: backend/lib/embeddings.ts
// Small embeddings client. Picks OpenAI (default), Voyage, or Cohere based on env.
// Output is always a 1024-dim float vector to match corpus_chunks.embedding (pgvector).

import { getOpenAI } from "./openai";

const PROVIDER = (process.env.EMBEDDINGS_PROVIDER ?? "openai").toLowerCase();

export async function embed(
  texts: string[],
  inputType: "document" | "query" = "document"
): Promise<number[][]> {
  if (texts.length === 0) return [];
  if (PROVIDER === "openai") return embedOpenAI(texts);
  if (PROVIDER === "voyage") return embedVoyage(texts, inputType);
  if (PROVIDER === "cohere") return embedCohere(texts, inputType);
  throw new Error(`Unknown EMBEDDINGS_PROVIDER: ${PROVIDER}`);
}

async function embedOpenAI(texts: string[]): Promise<number[][]> {
  const client = getOpenAI();
  const model = process.env.OPENAI_EMBEDDINGS_MODEL ?? "text-embedding-3-large";
  const r = await client.embeddings.create({
    model,
    input: texts,
    // Truncate to 1024 dims (text-embedding-3-* support the `dimensions` param via Matryoshka).
    dimensions: 1024,
  });
  return r.data.map((d) => d.embedding as number[]);
}

async function embedVoyage(texts: string[], inputType: "document" | "query"): Promise<number[][]> {
  const key = process.env.VOYAGE_API_KEY;
  if (!key) throw new Error("VOYAGE_API_KEY missing");
  const model = process.env.EMBEDDINGS_MODEL ?? "voyage-3-large";
  const r = await fetch("https://api.voyageai.com/v1/embeddings", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({ model, input: texts, input_type: inputType, output_dimension: 1024 }),
  });
  if (!r.ok) throw new Error(`voyage ${r.status}: ${await r.text()}`);
  const j = (await r.json()) as { data: { embedding: number[] }[] };
  return j.data.map((d) => d.embedding);
}

async function embedCohere(texts: string[], inputType: "document" | "query"): Promise<number[][]> {
  const key = process.env.COHERE_API_KEY;
  if (!key) throw new Error("COHERE_API_KEY missing");
  const model = process.env.EMBEDDINGS_MODEL ?? "embed-v3";
  const r = await fetch("https://api.cohere.com/v2/embed", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({
      model,
      texts,
      input_type: inputType === "document" ? "search_document" : "search_query",
      embedding_types: ["float"],
    }),
  });
  if (!r.ok) throw new Error(`cohere ${r.status}: ${await r.text()}`);
  const j = (await r.json()) as { embeddings: { float: number[][] } };
  return j.embeddings.float;
}

File: backend/railio/embeddings.py
"""Multi-provider embeddings (OpenAI / Voyage / Cohere). 1024-dim output."""

from __future__ import annotations

from typing import Literal

import httpx

from .config import get_settings
from .openai_client import get_openai

InputType = Literal["document", "query"]


async def embed(texts: list[str], input_type: InputType = "document") -> list[list[float]]:
    if not texts:
        return []
    provider = get_settings().embeddings_provider
    if provider == "openai":
        return await _embed_openai(texts)
    if provider == "voyage":
        return await _embed_voyage(texts, input_type)
    if provider == "cohere":
        return await _embed_cohere(texts, input_type)
    raise RuntimeError(f"Unknown EMBEDDINGS_PROVIDER: {provider}")


async def _embed_openai(texts: list[str]) -> list[list[float]]:
    client = get_openai()
    model = get_settings().openai_embeddings_model
    # text-embedding-3-* supports `dimensions` via Matryoshka.
    r = await client.embeddings.create(model=model, input=texts, dimensions=1024)
    return [list(d.embedding) for d in r.data]


async def _embed_voyage(texts: list[str], input_type: InputType) -> list[list[float]]:
    key = get_settings().voyage_api_key
    if not key:
        raise RuntimeError("VOYAGE_API_KEY missing")
    model = get_settings().openai_embeddings_model or "voyage-3-large"
    async with httpx.AsyncClient(timeout=60.0) as client:
        r = await client.post(
            "https://api.voyageai.com/v1/embeddings",
            headers={"Content-Type": "application/json", "Authorization": f"Bearer {key}"},
            json={
                "model": model,
                "input": texts,
                "input_type": input_type,
                "output_dimension": 1024,
            },
        )
        r.raise_for_status()
        data = r.json()["data"]
        return [list(d["embedding"]) for d in data]


async def _embed_cohere(texts: list[str], input_type: InputType) -> list[list[float]]:
    key = get_settings().cohere_api_key
    if not key:
        raise RuntimeError("COHERE_API_KEY missing")
    model = get_settings().openai_embeddings_model or "embed-v3"
    cohere_input_type = "search_document" if input_type == "document" else "search_query"
    async with httpx.AsyncClient(timeout=60.0) as client:
        r = await client.post(
            "https://api.cohere.com/v2/embed",
            headers={"Content-Type": "application/json", "Authorization": f"Bearer {key}"},
            json={
                "model": model,
                "texts": texts,
                "input_type": cohere_input_type,
                "embedding_types": ["float"],
            },
        )
        r.raise_for_status()
        return r.json()["embeddings"]["float"]
